framework/src/preconditioners/DualMortarPreconditioner.C

```diff
@@ -0,0 +1,588 @@
+        dofs.erase(*it_secondary);
+
+      std::vector<dof_id_type> vec_dofs(dofs.begin(), dofs.end());
+      _dof_sets_interior[vn].insert(make_pair(sd, vec_dofs));
+    }
+  }
+}
+
+void
+DualMortarPreconditioner::getDofVarInterface()
+{
+  _dof_sets_primary.resize(_n_vars);
+  _dof_sets_secondary.resize(_n_vars);
+  for (unsigned int vn = 0; vn < _n_vars; vn++)
+  {
+    // loop over boundary nodes
+    ConstBndNodeRange & range = * _mesh->getBoundaryNodeRange();
+    std::vector<dof_id_type> di;
+    for (const auto & bnode : range)
+    {
+      const Node * node_bdry = bnode->_node;
+      BoundaryID boundary_id = bnode->_bnd_id;
+
+      if (boundary_id == _secondary_boundary)
+      {
+          _dofmap->dof_indices(node_bdry, di, vn);
+          for (auto it = di.begin(); it!=di.end(); ++it)
+            if(std::find(_dof_sets_secondary[vn].begin(), _dof_sets_secondary[vn].end(), *it)==_dof_sets_secondary[vn].end())
+              _dof_sets_secondary[vn].push_back(*it);
+      }
+
+      if (boundary_id == _primary_boundary)
+      {
+        _dofmap->dof_indices(node_bdry, di, vn);
+        for (auto it = di.begin(); it!=di.end(); ++it)
+          if(std::find(_dof_sets_primary[vn].begin(), _dof_sets_primary[vn].end(), *it)==_dof_sets_primary[vn].end())
+            _dof_sets_primary[vn].push_back(*it);
+      }
+    }
+  }
+}
+
+void
+DualMortarPreconditioner::printDofSets()
+{
+  /// Dofs in the interior
+  for (auto it1 = _dof_sets_interior.begin(); it1 != _dof_sets_interior.end(); it1++)
+  {
+    std::cout << "variable "<<std::endl;
+    for (auto it2 = it1->begin(); it2 != it1->end(); it2++)
+    {
+      std::cout << " \tdofs in subdomain " << it2->first<<": ";
+      for (auto it3 = it2->second.begin(); it3 != it2->second.end(); it3++)
+        std::cout <<*it3 << " ";
+      std::cout <<"\n";
+    }
+  }
+
+  // primary
+  std::cout<<"primary dofs:\n";
+  for (auto it1=_dof_sets_primary.begin(); it1!=_dof_sets_primary.end(); it1++)
+  {
+    std::cout<< "\tvariable :";
+    for (auto it2=it1->begin(); it2!=it1->end(); it2++)
+      std::cout<<*it2<<" ";
+    std::cout<<std::endl;
+  }
+
+  // secondary
+  std::cout<<"secondary dofs:\n";
+  for (auto it1=_dof_sets_secondary.begin(); it1!=_dof_sets_secondary.end(); it1++)
+  {
+    std::cout<< "\tvariable :";
+    for (auto it2=it1->begin(); it2!=it1->end(); it2++)
+      std::cout<<*it2<<" ";
+    std::cout<<std::endl;
+  }
+}
+
+void
+DualMortarPreconditioner::condenseSystem()
+{
+  std::vector<dof_id_type> u1c=_dof_sets_primary[0];
+  std::vector<dof_id_type> u2c=_dof_sets_secondary[0];
+
+  std::vector<dof_id_type> lm=_dof_sets_secondary[1];
+
+  std::vector<dof_id_type> u1i=_dof_sets_interior[0][_primary_subdomain];
+  std::vector<dof_id_type> u2i=_dof_sets_interior[0][_secondary_subdomain];
+
+  _matrix->create_submatrix(* _D, lm, u2c);
+  _D->get_transpose(*_D);
```

Comment:
If we do `_matrix->create_submatrix(* _D, u2c, lm)`, we do not this transpose anymore

Reply:
Oh I remember here is a reason why I did not directly extract the sub matrix but choose to do the transpose (which is not optimal):

 The Jacobian is after applying BC, where the D block has some rows set to zero, similar with M. No BC is applied to the rows corresponding to the LMs, so I am extracting transpose(D), and transpose(M) first. 

I was thinking It would probably make more sense to do the condensing before applying the BC, but that probably will be really complex and does not bring much benefit

Reply:
The reason was basically the Jacobian (i.e., _matrix) is not symmetric anymore after applying the BC, where the entries of D and M are changed. However, since the rows associated with LM dofs are not touched, I am extracting D^T and M^T from there, and do the transpose to retrieve D and M

Reply:
Did you mean the interface may have an intersection with actual boundaries? And then a Dirichlet BCs will D and M because of some common mesh points between the interface and the boundaries.

Reply:
> Did you mean the interface may have an intersection with actual boundaries? And then a Dirichlet BCs will D and M because of some common mesh points between the interface and the boundaries.

Yes, you are right.